### DISCUSSION TITLE
Problem when center a window

### DISCUSSION THREAD

**Discussion by soldiereleven** [AvaloniaUI/Avalonia#17441] [Q&A] [answered]
I am trying to make the window at the top of the screen, as well at the center horizontally. I get the width of the window and caculated the position. However, in fact, the window is biased to the right.

Here's some codes:
```cs
ExtendClientAreaTitleBarHeightHint = 0;

// Set the window height as desired
Height = 70;

var screen = Screens.Primary;
if (screen != null)
{
    // Set the window width (e.g., 60% of screen width)
    Width = screen.Bounds.Width * 0.6;

    // Calculate the x-coordinate to center the window horizontally
    double centerX = 0.2 * screen.Bounds.Width;

    // Set the window position
    Position = new PixelPoint((int)centerX, 0);
}

// Ensure the window is positioned before showing
WindowStartupLocation = WindowStartupLocation.Manual;
```
and the displayed window:
![Screenshot 2024-11-06 222134](https://github.com/user-attachments/assets/159be8a7-8959-4afe-af87-202ae88b7673)

Any help is appreciated!

**Comment by thevortexcloud** [comment]
Generally the formula to centre things is:

`XPos = (ScreenWidth / 2) - (WindowWidth / 2)`

The idea being you want to place things halfway to the  end of the screen, and then you need to offset the position by half the window width so the window itself is centre aligned. Otherwise only the left edge of the window ends up centre aligned.

The formula for vertical centring is the same except using height instead of width.

I am not entirely clear what your formula is trying to do.

**Reply by soldiereleven** [reply]
First thanks for your suggestions. I'm sorry for not explain my question clearly and make you confused.

I found that the Position property of the window refers to the top left of the window to this position, so I multiply the screen width by 0.6 to get the width of the window, so that the left and right sides of the window have a width of 0.2 times the width of the screen, so that I can implement a floating bar that floats at the top of the screen.

I think that's right, because the top left corner of the window does line up with the vertical line in the center of the screen when I set the Position to (Width/2,0), but according to my code this hover bar doesn't fit nicely in the top center of the screen (it should be a hover bar in the top center of the screen like the default screenshot tool for Windows, maybe a little larger) 

@thevortexcloud I'm looking forward to your reply.

**Reply by soldiereleven** [reply]
![Screenshot 2024-11-06 222134](https://github.com/user-attachments/assets/d6110d26-c6ec-4946-9ed6-6086cf01a1a3)
It should be in the red rectangle.

**Reply by soldiereleven** [reply]
![image](https://github.com/user-attachments/assets/9bde1929-57f0-4073-a7b0-b0df512d69ab)
I find that if I set the scale to 100%, everything goes on well. Is the problem caused by scale or if I need to do a calculation on the DPI of the coordinates？

**Reply by thevortexcloud** [reply, marked as answer]
I am not too sure how Avalonia does DPI scaling as I have never used it on a screen where it mattered. I know a lot of things are supposed to be DPI independent though. 

But you can try multiplying the result/window position by `Scaling` property.

https://github.com/AvaloniaUI/Avalonia/blob/9cecb90ba1e681d1783e3a89db4b8d860859550f/src/Avalonia.Controls/Platform/Screen.cs#L60-L67

I imagine screen coordinates are not DPI scaled and window positions possibly are?

**Reply by soldiereleven** [reply]
Well, you are right. It seems that the window size is DPI scaled and the screen position is not. The reason why the window is skewed to the right is actually that the window is enlarged by 25%, so I divided the window width by DPI and it will display normally. But I still feel that the coordinates should be unified in relation to DPI scaling. And here's the final code:
```cs
var screen = Screens.Primary;
if (screen != null)
{
    double dpi = screen.Scaling;
    // Set the window width (e.g., 60% of screen width)
    Width = (screen.WorkingArea.Width * 0.6)/dpi;

    // Calculate the x-coordinate to center the window horizontally
    double centerX = screen.WorkingArea.Width * 0.2;
    //Trace.WriteLine(centerX);
    //Trace.WriteLine(screen.WorkingArea.Width);
    // Set the window position
    Position = new PixelPoint((int)Math.Round(centerX), 0);
    
}

// Ensure the window is positioned before showing
WindowStartupLocation = WindowStartupLocation.Manual;
```

**Reply by thevortexcloud** [reply]
I am not sure it makes sense for _physical_ screen coordinates to be scaled (after all, 1 physical pixel is always 1 pixel wide). Whereas the window is something hosted in the screen and needs to know how many physical pixels to occupy. That said I think it would make sense for there to be a method to automatically scale something to the DPI scaled coordinates.

**Comment by theoractice** [comment]
#19717 might be helpful. BTW I tried to use `SetWindowPos` to set the size of my Avalonia/WPF app and it worked perfectly.
Simply ignore the `Width/Height` thing when your primary concern is pixel position rather than DPI scaling.
